db/sqlinterfaces.c

```diff
@@ -752,57 +746,42 @@ int newsql_write_response(struct sqlclntstate *clnt, int type,
     hdr.length = ntohl(len);
 
     rc = pthread_mutex_lock(&clnt->write_lock);
-
     if (rc != 0) {
         logmsg(LOGMSG_FATAL, "couldnt get clnt->write_lock\n");
         exit(1);
     }
 
-    rc = sbuf2write((char *)&hdr, sizeof(struct newsqlheader), sb);
-    if (rc != sizeof(struct newsqlheader)) {
-        rc = pthread_mutex_unlock(&clnt->write_lock);
-        if (rc != 0) {
-            logmsg(LOGMSG_FATAL, "couldnt put clnt->write_lock\n");
-            exit(1);
-        }
-        free(dta);
-        return -1;
+    SBUF2 *sb = clnt->sb;
+    int wlen = sbuf2write((char *)&hdr, sizeof(struct newsqlheader), sb);
+    if (wlen != sizeof(struct newsqlheader)) {
+        rc = -1;
+        goto done;
     }
 
     if (dta) {
-        rc = sbuf2write(dta, len, sb);
-        if (rc != len) {
-
-            if (gbl_dump_fsql_response) {
-                logmsg(LOGMSG_USER, "sbuf2write error for %s rc=%d\n", clnt->sql,
-                        rc);
-            }
-
-            rc = pthread_mutex_unlock(&clnt->write_lock);
-            if (rc != 0) {
-                logmsg(LOGMSG_FATAL, "couldnt get clnt->write_lock\n");
-                exit(1);
-            }
-
-            free(dta);
-            return -1;
+        wlen = sbuf2write(dta, len, sb);
+        if (wlen != len) {
+            if (gbl_dump_fsql_response)
+                logmsg(LOGMSG_USER, "sbuf2write error for %s wlen=%d\n",
+                       clnt->sql, wlen);
+            rc = -1;
+            goto done;
         }
     }
 
-    if (flush) {
+    if (flush)
         sbuf2flush(sb);
-    }
 
-    rc = pthread_mutex_unlock(&clnt->write_lock);
-    if (rc != 0) {
-        logmsg(LOGMSG_FATAL, "couldnt get clnt->write_lock\n");
+done:
+    if (pthread_mutex_unlock(&clnt->write_lock) != 0) {
+        logmsg(LOGMSG_FATAL, "couldnt put clnt->write_lock\n");
         exit(1);
```

Comment:
UNRELATED: Just a thought, logmsg() should trigger exit(1)/abort() on LOGMSG_FATAL.

Reply:
Hmm, that does not sound bad. I wonder if there is any circumstance where we don't exit after logmsg_fatal.

Reply:
> I wonder if there is any circumstance where we don't exit after logmsg_fatal.

In such cases it shouldn't be fatal.